### PR TITLE
fix(core.crypto): UTF characters wrongly encoded/decoded with AES

### PR DESCRIPTION
**Pull request by salvatore-coppola**
Different fix for the problem in #4481 but in this case we create encoders and decoders capable of raise exceptions to detect invalid UTF-8 characters and try to fallback to the default charset.

### CODE REVIEW

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java [hunk 1]
```diff
@@ -89,21 +120,50 @@ public char[] encryptAes(char[] value) throws KuraException {
...
-        } catch (InvalidKeyException | IllegalBlockSizeException | BadPaddingException e) {
+        } catch (InvalidKeyException | IllegalBlockSizeException | BadPaddingException | CharacterCodingException e) {
             throw new KuraException(KuraErrorCode.ENCODE_ERROR, VALUE_EXCEPTION_CAUSE);
         } catch (InvalidAlgorithmParameterException e) {
             throw new KuraException(KuraErrorCode.ENCODE_ERROR, PARAMETER_EXCEPTION_CAUSE);
         }
 
     }
 
+    private byte[] charArrayToByteArray(char[] value) throws CharacterCodingException {
```
Comment (MMaiero) [inline kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java:113]: Is this needed?
Can the encoding to utf8 fail due to CharacterCodingException?

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java [hunk 2]
```diff
@@ -64,6 +71,30 @@ public class CryptoServiceImpl implements CryptoService {
     private final SecureRandom random = new SecureRandom();
     private SystemService systemService;
 
+    private CharsetEncoder utf8Encoder;
```
Comment (MMaiero) [inline kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: Please see that CharsetEncoder is not thread-safe: https://docs.oracle.com/javase/7/docs/api/java/nio/charset/CharsetEncoder.html

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java [hunk 3]
```diff
@@ -64,6 +71,30 @@ public class CryptoServiceImpl implements CryptoService {
...
+    private CharsetDecoder utf8Decoder;
```
Comment (MMaiero) [inline kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: Same for decoder

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java [hunk 4]
```diff
@@ -89,21 +120,50 @@ public char[] encryptAes(char[] value) throws KuraException {
...
     }
 
+    private byte[] charArrayToByteArray(char[] value) throws CharacterCodingException {
+
+        ByteBuffer byteBuffer;
+        try {
+            byteBuffer = this.utf8Encoder.encode(CharBuffer.wrap(value));
+        } catch (CharacterCodingException e) {
+            // fallback for backward compatibility
+            byteBuffer = this.platformEncoder.encode(CharBuffer.wrap(value));
```
Comment (MMaiero) [inline kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: Does the behaviour here change compared to the past?
The encoder is set to CodingErrorAction.REPORT so it fails on issues on encoding.
But in the current code it acts silently replacing characters. So, to keep backward compatibility, we should use the old code

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java [hunk 5]
```diff
@@ -89,21 +120,50 @@ public char[] encryptAes(char[] value) throws KuraException {
...
+        return encodedBytes;
+    }
+
+    private char[] byteArrayToCharArray(byte[] value) throws CharacterCodingException {
+        CharBuffer charBuffer;
+        try {
+            charBuffer = this.utf8Decoder.decode(ByteBuffer.wrap(value));
+        } catch (CharacterCodingException e) {
+            // fallback for backward compatibility
+            charBuffer = this.platformDecoder.decode(ByteBuffer.wrap(value));
```
Comment (MMaiero) [inline kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: Same as for the previous comment:

> Does the behaviour here change compared to the past?
The encoder is set to CodingErrorAction.REPORT so it fails on issues on encoding.
But in the current code it acts silently replacing characters. So, to keep backward compatibility, we should use the old code

We should probably use the old code to keep the current behaviour

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java (hunk 1 above)
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java:113]: the encoder has been instruced to thrown the exception in the CryptoServiceImpl contructor.

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java (hunk 2 above)
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: i'll synchronize the methods using them.

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java (hunk 4 above)
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: I'll change the CodingErrorAction with replace for the platform encoder/decoder

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java (hunk 2 above)
Comment (nicolatimeus) [reply kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: Is it better to make the methods synchronized or using a dedicated encoder/decoder for each encrypt/decrypt operation?

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java (hunk 4 above)
Comment (nicolatimeus) [reply kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: The Javadoc of [getBytes()](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#getBytes()) says that "The behavior of this method when this string cannot be encoded in the default charset is unspecified.", so it is not given that switching to REPLACE will bring back the old behavior.. probably we should switch back to `getBytes()` here.

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java (hunk 5 above)
Comment (nicolatimeus) [reply kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: We should probably go back to `new String(value)` also here

File: kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java (hunk 2 above)
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: > Is it better to make the methods synchronized or using a dedicated encoder/decoder for each encrypt/decrypt operation?

I don't know, it seems too expensive to me to create an encoder/decoder every time.

Comment (nicolatimeus) [reply kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java]: It doesn't seem that expensive looking at openjdk impl https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/sun/nio/cs/UTF_8.java#L57 https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/java/nio/charset/Charset-X-Coder.java.template

**Review by nicolatimeus — APPROVED** [review]

**Pull request merged**